src/inmanta/data/__init__.py

```diff
@@ -4439,6 +4509,76 @@ async def trim(cls, environment: UUID, connection: Optional[Connection] = None)
             connection=connection,
         )
 
+    @classmethod
+    async def populate_for_version(
+        cls, environment: uuid.UUID, model_version: int, connection: Optional[Connection] = None
+    ) -> None:
+        """
+        Make sure that the resource_persistent_state table has a record for each resource present in the
+        given model version. This method assumes that the given model_version is the latest released version.
+        """
+        await cls._execute_query(
+            f"""
+            INSERT INTO {cls.table_name()} (
+                environment,
+                resource_id,
+                resource_type,
+                agent,
+                resource_id_value,
+                current_intent_attribute_hash,
+                is_undefined,
+                is_orphan,
+                deployment_result,
+                blocked_status
+            )
+            SELECT
+                r.environment,
+                r.resource_id,
+                r.resource_type,
+                r.agent,
+                r.resource_id_value,
+                r.attribute_hash,
+                r.status = 'undefined'::public.resourcestate,
+                FALSE,
+                'NEW',
+                CASE
+                    WHEN
+                        r.status = 'undefined'::public.resourcestate
+                        OR r.status = 'skipped_for_undefined'::public.resourcestate
+                    THEN 'YES'
+                    ELSE 'NO'
+                END
+            FROM {Resource.table_name()} AS r
+            WHERE r.environment=$1 AND r.model=$2 AND NOT EXISTS(
+                SELECT *
+                FROM {cls.table_name()} AS rps
+                WHERE rps.environment=r.environment AND rps.resource_id=r.resource_id
+            )
+            """,
+            environment,
+            model_version,
+            connection=connection,
+        )
+
+    def get_compliance_status(self) -> state.ComplianceStatus:
+        """
+        Return the ComplianceStatus associated with this resource_persistent_state.
+        """
+        if self.is_orphan:
+            return state.ComplianceStatus.ORPHAN
+        elif self.is_undefined:
+            return state.ComplianceStatus.UNDEFINED
+        elif self.blocked_status is state.BlockedStatus.YES:
+            return state.ComplianceStatus.NON_COMPLIANT
```

Comment:
Semantically, I'd say not. Implementation-wise, I'm not sure, I need to check. But if it's possible, I'd consider that a bug in the scheduler rather than in this function.

Reply:
I agree with that

Reply:
> `BLOCKED <=> undefined`

This is not correct though (not in the current implementation, where `ComplianceStatus.UNDEFINED <=> intent undefined`). A resource can be blocked because *a dependency* is undefined.

Which makes me realize that I was missing that same case in my reply above. I think Wouter may be right that a resource can be both blocked and compliant, namely if it was in a good state, hasn't had any changes to its attributes since, and a dependency got into an undefined state since then. It then all boils down to what we consider `ComplianceStatus.UNDEFINED` to mean:
- if it means `ComplianceStatus.UNDEFINED <=> intent undefined`, then we can indeed be both blocked and compliant
- if on the other hand it means `ComplianceStatus.UNDEFINED <=> we don't know enough to judge compliance`, I think we should end up in that state when we get blocked, because we can't execute incremental deploys anymore, therefore our compliance data will go stale. In this case I think that we (@wouterdb perhaps in your PR?) should set the compliance status to undefined when we transitively block a resource. In this case this function is correct, as is Arnaud's latest statement.

I think the second definition makes a lot of sense, though we might consider changing the name of the `ComplianceStatus.UNDEFINED` value in that case, because it could be confusing that it is the same name we use for undefined intent, while meaning something broader.

Wdyt?

Reply:
> Why would another resources undefinedness cause uncertainty over our compliance?

Mainly because we lose the ability to verify it, so the compliance confidence will go down with time.

> we assume compliance remains unless we have a failure/skip or intent is updated. It is the fundamental assumption underlying increments.

Not entirely though: when a deploy fails we assume non-compliance even if it was compliant before, while this could mean all sort of things. So I would say the line is a bit more blurred than that.

> 2. ...

I'm not sure I follow 2, but I'm not sure it's worth getting into.

Overall I can get behind the first definition (we should probably document it). In which case this function needs to be updated accordingly.